### PR TITLE
Separate "indexed objects" by type

### PR DESCRIPTION
**Pull request by erikbern**
We have a few `Dict[str, _Object]` where the object is either a function or a class. It makes sense to separate these by type. Starting with `App.indexed_objects`. I'll do `Running_app.tag_to_object_id` later.

### CODE REVIEW

File: modal/app.py [hunk 1]
```diff
@@ -507,7 +517,11 @@ def registered_entrypoints(self) -> Dict[str, _LocalEntrypoint]:
 
     @property
     def indexed_objects(self) -> Dict[str, _Object]:
-        return self._indexed_objects
+        deprecation_warning(
+            (2024, 11, 25),
+            "`app.indexed_objects` is deprecated! Use `app.registered_functions` or `app.registered_classes` instead.",
```
Comment (mwaskom) [inline modal/app.py:522]: I think we might be planning on deprecating those too? 😁

**Review by mwaskom — APPROVED** [review]
:+1: we have some very annoying code to unmix the `tag_to_object_ids` into the separate `function_ids` / `class_ids` for the `AppPublishRequest`, will be nice to clean that up!

File: modal/app.py [hunk 2]
```diff
@@ -177,7 +177,8 @@ def foo():
 
     _name: Optional[str]
     _description: Optional[str]
-    _indexed_objects: Dict[str, _Object]
+    _functions: Dict[str, _Function]
+    _classes: Dict[str, _Cls]
```
Comment (mwaskom) [inline modal/app.py:181]: `classes`? Or ... `clse`? 🤔 😭

Comment (erikbern) [reply modal/app.py:181]: yeah idk, no strong feelings

**Pull request merged**